antea/scripts/process_mc_petit_FBK.py

```diff
@@ -0,0 +1,307 @@
+    v_frac    = 1 - exp(diff_time, rec_time)
+    charges   = np.insert(v_frac, 0, 1)
+    df.insert(len(df.columns), 'charge', charges.astype(float))
+
+    return df
+    
+
+def process_mc_petit_FBK(input_file: str, output_file: str, recovery_time: int):
+
+    """
+    This function selects data events in coincidence and it saves the maximum
+    charge of the event in each plane and the minimum time. It also computes
+    the charge obtained in a SiPM with and without saturation. Finally, it sums 
+    the charge of 4 in 4 sensors to compare with real data.
+    """
+    
+    DataSiPM_pb     = db.DataSiPM('petalo', 12334, 'PB')
+    DataSiPM_pb_idx = DataSiPM_pb.set_index('SensorID')
+
+    ### parameters for single photoelectron convolution in SiPM response
+    tau_sipm       = [100, 15000]
+    time_window    = 5000
+    time           = np.arange(0, time_window)
+    spe_resp, norm = shf.normalize_sipm_shaping(time, tau_sipm)
+
+    n_pe          = 1  # number of first photoelectron to be considered for times
+    sigma_sipm    = 40 #ps SiPM jitter
+    sigma_elec    = 30 #ps electronic jitter
+    timestamp_thr = 0.25 # PETsys threshold to extract the timestamp
+    
+    ### Data to save
+    sensor_id0, sensor_id2   = [], []
+    max_charge0, max_charge2 = [], []
+    max_or0, max_or2         = [], []
+    tmin0, tmin2             = [], []
+    tmin_ch0, tmin_ch2       = [], []
+    tmin_sid0, tmin_sid2     = [], []
+    event_ids                = []
+
+    
+    try:
+        tof_response = load_mcTOFsns_response(input_file)
+    except ValueError:
+        print(f'File {input_file} not found')
+        exit()
+    except OSError:
+        print(f'File {input_file} not found')
+        exit()
+    except KeyError:
+        print(f'No object named MC/tof_sns_response in file {input_file}')
+        exit()
+    except:
+        print(f'Unknown error in {input_file}')
+        exit()
+    print(f'Analyzing file {input_file}')
+
+
+    ### Add column for real sensor_id instead of cells_id          
+    tof_response = tof_response.rename(columns={'sensor_id':'sns_cells'})
+    tof_sensors  = tof_response.sns_cells.values // 10000
+    tof_response['sensor_id'] = tof_sensors
+    tof_response = num_sensors_4in4(tof_response)
+
+    ### Apply saturation:
+    events = tof_response.event_id.unique()
+    n_evts           = len(events)
+    n_evts_per_bunch = int(100)
+    n_bunches        = int(np.ceil(n_evts / n_evts_per_bunch))
+    print(n_evts, n_bunches)
+    for n in range(n_bunches):
+        print('bunch', n)
+        evt_range = events[n*n_evts_per_bunch:(n+1)*n_evts_per_bunch]
+        b_sns   = tof_response[tof_response.event_id.isin(evt_range)]
+        charge_df = b_sns.groupby(['event_id','sns_cells'], as_index=False).apply(sim_saturation_vicente, rec_time=recovery_time)
+
+        ### Create DataFrame for charge data, with and without saturation:
+        print('creating charge dataFrames')
```

Comment:
These printouts can be useful in debugging, but I think they can be removed in the official code.